### PR TITLE
Allow generate and remote_state to be set as attrs

### PR DESCRIPTION
**Pull request by brikis98**
This allows us to configure code generation and remote state dynamically. See the documentation updates for examples.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -777,7 +826,26 @@ func convertToTerragruntConfig(
 		terragruntConfig.IamRole = *terragruntConfigFromFile.IamRole
 	}
 
-	for _, block := range terragruntConfigFromFile.GenerateBlocks {
+	generateBlocks := []terragruntGenerateBlock{}
+	generateBlocks = append(generateBlocks, terragruntConfigFromFile.GenerateBlocks...)
+
+	if terragruntConfigFromFile.GenerateAttrs != nil {
+		remoteStateMap, err := parseCtyValueToMap(*terragruntConfigFromFile.GenerateAttrs)
```
Comment (yorinasub17) [inline config/config.go:833]:
```suggestion
		generateMap, err := parseCtyValueToMap(*terragruntConfigFromFile.GenerateAttrs)
```

File: config/config.go [hunk 2]
```diff
@@ -777,7 +826,26 @@ func convertToTerragruntConfig(
...
+		if err != nil {
+			return nil, err
+		}
+
+		for name, block := range remoteStateMap {
```
Comment (yorinasub17) [inline config/config.go:838]:
```suggestion
		for name, block := range generateMap {
```

File: docs/_docs/04_reference/config-blocks-and-attributes.md [hunk 3]
```diff
@@ -225,6 +225,34 @@ The `remote_state` block supports the following arguments:
...
+  remote_state = {
+    backend = "s3"
+    config = {
+      bucket = "mybucket"
+      key    = "path/to/my/key"
+      region = "us-east-1"
+    }
+  }
+}
+```
```
Comment (yorinasub17) [inline docs/_docs/04_reference/config-blocks-and-attributes.md:242]: Side note: It isn't strictly necessary to set this as a local, as `read_terragrunt_config` will expose the `remote_state` block under the `remote_state` attribute.

**Review by yorinasub17 — COMMENTED** [review]
LGTM! Only had some style nit comments

File: config/config.go (hunk 1 above)
Comment (brikis98) [reply config/config.go]: Whoops, good catch. Fixed in https://github.com/gruntwork-io/terragrunt/pull/1639/commits/b6f61c8300610841e5f4ae4b466f5d76490c0374.

File: config/config.go (hunk 2 above)
Comment (brikis98) [reply config/config.go]: Whoops, good catch. Fixed in https://github.com/gruntwork-io/terragrunt/pull/1639/commits/b6f61c8300610841e5f4ae4b466f5d76490c0374.

File: docs/_docs/04_reference/config-blocks-and-attributes.md (hunk 3 above)
Comment (brikis98) [reply docs/_docs/04_reference/config-blocks-and-attributes.md:242]: Oh, interesting!

What about `generate` blocks? Is there a way to access those from what `read_terragrunt_config` returns?

Comment (yorinasub17) [reply docs/_docs/04_reference/config-blocks-and-attributes.md:242]: Generate is also available under `generate` attribute as a map. See https://github.com/gruntwork-io/terragrunt/blob/master/test/integration_test.go#L3043

**Review by yorinasub17 — APPROVED** [review]

File: docs/_docs/04_reference/config-blocks-and-attributes.md (hunk 3 above)
Comment (brikis98) [reply docs/_docs/04_reference/config-blocks-and-attributes.md:240]: Oh, very handy. OK, docs updated: https://github.com/gruntwork-io/terragrunt/pull/1639/commits/675fc22446876104c2ccab49523c1c6e4020d10c.

**Comment by brikis98** [comment]
Thx for the reviews! Could I get one more ship it?

**Review by yorinasub17 — APPROVED** [review]
LGTM! I didn't think the native formats were better, but then I realized you get type checking that way so it's good! 👍

**Comment by brikis98** [comment]
> LGTM! I didn't think the native formats were better, but then I realized you get type checking that way so it's good! 👍

Yea, it's both cleaner to read, and safer. Nice call.

Thx for the reviews. Merging now.

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/terragrunt/releases/tag/v0.28.21

**Pull request merged**